dandischema/digests/dandietag.py

```diff
@@ -0,0 +1,202 @@
+        return self._part_gen[number]
+
+    def get_parts(self) -> Iterator[Part]:
+        return iter(self._part_gen)
+
+    def get_next_part(self) -> Optional[Part]:
+        if self._next_index < self.part_qty:
+            return self._part_gen[self._next_index + 1]
+        else:
+            return None
+
+    def get_part_etag(self, p: Part) -> Optional[str]:
+        return self._md5_digests[p.number - 1].hex()
+
+    def as_str(self) -> str:
+        if not self.complete:
+            raise ValueError("Not all part hashes submitted")
+        blob = b""
+        for d in self._md5_digests:
+            assert d is not None
+            blob += d
+        parts_digest = md5(blob).hexdigest()
+        return f"{parts_digest}-{len(self._md5_digests)}"
+
+    @classmethod
+    def from_file(
+        cls, path: Union[str, bytes, "os.PathLike[str]", "os.PathLike[bytes]"]
+    ) -> "DandiETag":
+        etag = cls(file_size=os.path.getsize(path))
+        with open(path, "rb") as f:
+            for part in etag.get_parts():
+                etag.update(f.read(part.size))
+        return etag
+
+    def _add_digest(self, p: Part, part_digest: bytes) -> None:
+        i = p.number - 1
+        if self._md5_digests[i] is not None:
+            raise RuntimeError(f"Digest for part {p.number} submitted more than once")
+        self._md5_digests[i] = part_digest
+        self._update_index()
+
+    def _add_next_digest(self, part_digest: bytes) -> None:
+        if self.complete:
+            raise RuntimeError(
+                "Trying to update DandiETag with a new digest having already"
+                f" processed all {self.part_qty} parts"
+            )
+        self._md5_digests[self._next_index] = part_digest
+        self._update_index()
+
+    def _update_index(self) -> None:
+        while (
+            self._next_index < self.part_qty
+            and self._md5_digests[self._next_index] is not None
+        ):
+            self._next_index += 1
+
+    def update(self, block: bytes, part: Optional[Part] = None) -> None:
+        """Update etag with the new block of data"""
+        if self._partial_blob:
+            raise ValueError("Digesting new part when current part is not complete")
+        part_digest = md5(block).digest()
+        if part is None:
+            self._add_next_digest(part_digest)
+        else:
+            self._add_digest(part, part_digest)
+
+    def partial_update(self, block: bytes) -> None:
+        self._partial_blob += block
+        p = self.get_next_part()
+        while p is not None and p.size <= len(self._partial_blob):
+            next_block = self._partial_blob[: p.size]
+            self._partial_blob = self._partial_blob[p.size :]
+            self._add_next_digest(md5(next_block).digest())
+            p = self.get_next_part()
+        if p is None and self._partial_blob:
+            raise ValueError("Partial update extended past end of file")
+
+
+class ETagHashlike:
```

Comment:
this class (and used `.partial_update`) seems to lack tests code coverage -- we never had it for this portion or it was triggered by dandi-cli other functionality somehow?  should be easy to add a dedicated unittest to ensure a more complete test coverage

Reply:
Test for `ETagHashlike` added.